libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,400 @@
+                         std::forward<_ArgTypes>(__args)...);
+    }
+  };
+
+  template <class _Functor>
+  using __function_wrappers = __function_wrappers_impl<remove_cvref_t<_Functor>>;
+
+  template <class _VT>
+  consteval static bool __is_callable_from_impl() {
+    if (_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) {
+      return is_nothrow_invocable_r_v< _ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF, _ArgTypes...> &&
+             is_nothrow_invocable_r_v<_ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS, _ArgTypes...>;
+    } else {
+      return is_invocable_r_v<_ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_CV_REF, _ArgTypes...> &&
+             is_invocable_r_v<_ReturnT, _VT _LIBCPP_MOVE_ONLY_FUNCTION_INV_QUALS, _ArgTypes...>;
+    }
+    return false;
+  }
+
+  template <class _VT>
+  static constexpr bool __is_callable_from = __is_callable_from_impl<decay_t<_VT>>();
+
+  template <class _Func, class... _Args>
+  _LIBCPP_HIDE_FROM_ABI void __construct(_Args&&... __args) {
+    static_assert(is_constructible_v<decay_t<_Func>, _Func>);
+
+    using _FuncWraps = __function_wrappers<_Func>;
+    using _UnRefFunc = remove_reference_t<_Func>;
+
+    if constexpr (_Store::__fits_in_trivially_destructible_buffer<_Func>) {
+      auto& __data   = __storage_.__trivially_destructible_;
+      __data.__call_ = std::bit_cast<void*>(&_FuncWraps::__call);
+      std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__data_), std::forward<_Args>(__args)...);
+      __data.__status_ = _Store::_Status::_TriviallyDestructible;
+    } else if constexpr (_Store::__fits_in_trivially_relocatable_buffer<_Func>) {
+      auto& __data      = __storage_.__trivially_relocatable_;
+      __data.__call_    = std::bit_cast<void*>(&_FuncWraps::__call);
+      __data.__destroy_ = &_FuncWraps::__destroy;
+      std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__data_), std::forward<_Args>(__args)...);
+      __data.__status_ = _Store::_Status::_TriviallyRelocatable;
+    } else {
+      auto& __data      = __storage_.__large_;
+      __data.__call_    = std::bit_cast<void*>(&_FuncWraps::__call);
+      __data.__destroy_ = &_FuncWraps::__destroy;
+      __data.__heap_    = static_cast<std::byte*>(::operator new(sizeof(_Func), std::align_val_t(alignof(_Func))));
+      std::construct_at(reinterpret_cast<_UnRefFunc*>(__data.__heap_), std::forward<_Args>(__args)...);
+      __data.__status_ = _Store::_Status::_Heap;
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI void __reset() {
+    using enum _Store::_Status;
+    switch (__storage_.__large_.__status_) {
+    case _NotEngaged:
+    case _TriviallyDestructible:
+      break;
+
+    case _TriviallyRelocatable:
+      __storage_.__trivially_relocatable_.__destroy_(&__storage_.__trivially_relocatable_.__data_);
+      break;
+    case _Heap:
+      __storage_.__large_.__destroy_(__storage_.__large_.__heap_);
+      ::operator delete(__storage_.__large_.__heap_);
+      break;
+    }
+    __storage_.__large_.__status_ = _NotEngaged;
```

Comment:
If you make my suggested changes above, to the lambdas assigned in the `_Heap` case, then your `case _TriviallyRelocatable` and `case _Heap` here can be collapsed together.
Then, you have only two cases to deal with: (1) There is an object here that needs destroying via `__destroy_(__heap_)`, or (2) there is no object here that needs destroying (either it's trivially dtable, or it's disengaged). We can represent that second case easily by setting `__destroy_ == nullptr`. Then this entire function becomes
```
_LIBCPP_HIDE_FROM_ABI void __reset() {
  if (__storage_.__destroy_ != nullptr) {
    __storage_.__destroy_(__storage_.__data_);
  }
```
And we can do the same mechanical transformation for `__call_`. And then (unless I'm mistaken) we don't need `__status_` anymore, and we don't need a union of `__trivially_relocatable_` and `__destroy_` and so on anymore, either. Things get much simpler.

I think there is one downside to my idea, though: IIUC, right now your layout for trivially relocatable types is (call, destroy, 31 bytes of data, status) and your layout for trivially destructible types is (call, 39 bytes of data, status). So you can store a trivially destructible type of up to 39 bytes. Whereas I'm proposing a uniform layout of (call, destroy, 32 bytes of data). Which _increases_ your ability to store trivially relocatable types, but _decreases_ your ability to store trivially destructible types (from 39 bytes down to 32).

One more idea is to do what libc++'s `std::any` does, and combine (call, destroy) into a single function pointer. See https://www.youtube.com/watch?v=tbUCHifyT24&ab_channel=CppCon&t=24m48s for an example. Then you can use all 40 bytes for both trivially-dtable and trivially-relocable types. (Or, you can reclaim 8 of those bytes and shrink the total size of `std::move_only_function`.)